### **1. load_digits : 손글씨를 분류해 봅시다**

**(1) 필요한 모듈 import하기**

In [347]:
import sklearn

print(sklearn.__version__)

1.0.2


In [348]:
from sklearn.datasets import load_digits
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report

In [349]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn import svm
from sklearn.linear_model import SGDClassifier
from sklearn.linear_model import LogisticRegression

**(2) 데이터 준비**

In [350]:
digits = load_digits()

**(3) 데이터 이해하기**



*   Feature Data 지정하기



In [351]:
digits_feature = digits.data



*   Label Data 지정하기



In [352]:
digits_label = digits.target



*   Target Names 출력해 보기



In [353]:
digits.target_names

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])



*   데이터 Describe 해 보기



In [354]:
print(digits.DESCR)

.. _digits_dataset:

Optical recognition of handwritten digits dataset
--------------------------------------------------

**Data Set Characteristics:**

    :Number of Instances: 1797
    :Number of Attributes: 64
    :Attribute Information: 8x8 image of integer pixels in the range 0..16.
    :Missing Attribute Values: None
    :Creator: E. Alpaydin (alpaydin '@' boun.edu.tr)
    :Date: July; 1998

This is a copy of the test set of the UCI ML hand-written digits datasets
https://archive.ics.uci.edu/ml/datasets/Optical+Recognition+of+Handwritten+Digits

The data set contains images of hand-written digits: 10 classes where
each class refers to a digit.

Preprocessing programs made available by NIST were used to extract
normalized bitmaps of handwritten digits from a preprinted form. From a
total of 43 people, 30 contributed to the training set and different 13
to the test set. 32x32 bitmaps are divided into nonoverlapping blocks of
4x4 and the number of on pixels are counted in each blo

**(4) train, test 데이터 분리**

In [355]:
X_train, X_test, y_train, y_test = train_test_split(digits_feature, digits_label, test_size=0.2, random_state=1)

**(5) 다양한 모델로 학습시켜보기**



*   Decision Tree 사용해 보기



In [356]:
decision_tree = DecisionTreeClassifier(random_state=1)
decision_tree.fit(X_train, y_train)
y_pred_decision_tree = decision_tree.predict(X_test)

print(classification_report(y_test, y_pred_decision_tree))

              precision    recall  f1-score   support

           0       1.00      0.93      0.96        43
           1       0.91      0.83      0.87        35
           2       0.83      0.83      0.83        36
           3       0.78      0.71      0.74        41
           4       0.82      0.84      0.83        38
           5       0.78      0.97      0.87        30
           6       0.88      0.97      0.92        37
           7       0.78      0.76      0.77        37
           8       0.83      0.86      0.85        29
           9       0.75      0.71      0.73        34

    accuracy                           0.84       360
   macro avg       0.84      0.84      0.84       360
weighted avg       0.84      0.84      0.84       360




*   Random Forest 사용해 보기

In [357]:
random_forest = RandomForestClassifier(random_state=2)
random_forest.fit(X_train, y_train)
y_pred_random_forest = random_forest.predict(X_test)

print(classification_report(y_test, y_pred_random_forest))

              precision    recall  f1-score   support

           0       1.00      0.95      0.98        43
           1       1.00      1.00      1.00        35
           2       1.00      1.00      1.00        36
           3       1.00      0.98      0.99        41
           4       0.95      1.00      0.97        38
           5       0.97      1.00      0.98        30
           6       1.00      1.00      1.00        37
           7       1.00      0.97      0.99        37
           8       0.97      0.97      0.97        29
           9       0.94      0.97      0.96        34

    accuracy                           0.98       360
   macro avg       0.98      0.98      0.98       360
weighted avg       0.98      0.98      0.98       360





* SVM 사용해 보기



In [358]:
svm = svm.SVC()
svm.fit(X_train, y_train)
y_pred_svm = svm.predict(X_test)

print(classification_report(y_test, y_pred_svm))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        43
           1       1.00      1.00      1.00        35
           2       1.00      1.00      1.00        36
           3       1.00      1.00      1.00        41
           4       1.00      1.00      1.00        38
           5       0.97      1.00      0.98        30
           6       1.00      1.00      1.00        37
           7       1.00      0.97      0.99        37
           8       1.00      0.97      0.98        29
           9       0.94      0.97      0.96        34

    accuracy                           0.99       360
   macro avg       0.99      0.99      0.99       360
weighted avg       0.99      0.99      0.99       360




*   SGD Classifier 사용해 보기

In [359]:
sgd = SGDClassifier()
sgd.fit(X_train, y_train)
y_pred_sgd = sgd.predict(X_test)

print(classification_report(y_test, y_pred_sgd))

              precision    recall  f1-score   support

           0       1.00      0.98      0.99        43
           1       0.94      0.94      0.94        35
           2       1.00      0.97      0.99        36
           3       1.00      0.90      0.95        41
           4       0.90      1.00      0.95        38
           5       0.97      0.93      0.95        30
           6       0.97      1.00      0.99        37
           7       1.00      0.92      0.96        37
           8       0.83      1.00      0.91        29
           9       0.91      0.88      0.90        34

    accuracy                           0.95       360
   macro avg       0.95      0.95      0.95       360
weighted avg       0.96      0.95      0.95       360





*   Logistic Regression 사용해 보기



In [360]:
logistic_regression = LogisticRegression(max_iter=3000)
logistic_regression.fit(X_train, y_train)
y_pred_logistic_regression = logistic_regression.predict(X_test)

print(classification_report(y_test, y_pred_logistic_regression))

              precision    recall  f1-score   support

           0       1.00      0.98      0.99        43
           1       0.97      1.00      0.99        35
           2       1.00      1.00      1.00        36
           3       0.98      0.98      0.98        41
           4       0.95      1.00      0.97        38
           5       0.88      0.93      0.90        30
           6       1.00      1.00      1.00        37
           7       1.00      0.95      0.97        37
           8       0.93      0.93      0.93        29
           9       1.00      0.94      0.97        34

    accuracy                           0.97       360
   macro avg       0.97      0.97      0.97       360
weighted avg       0.97      0.97      0.97       360



**(6) 모델을 평가해 보기**

손글씨 데이터라는 점을 주목했을 때,

*precision*과 *recall*을 구분하는 것이 큰 의미가 없다.

손글씨 '1'을 예로 들면,

> 1) 실제 '1'을 다른 숫자로 예측한 것이 많을 경우(= *recall* 값이 작아지는 경우)

> 2) 실제 '1'이 아닌 숫자를 '1'로 예측한 것이 많을 경우(= *precision* 값이 작아지는 경우)는

똑같은 상황으로 간주할 수 있다. 손글씨마다 특정한 의미가 존재한다고 볼 수 없기 때문이다.

즉, 이는 단순하게 손글씨의 실제 값과 예측 값이 일치하느냐 일치하지 않느냐의 문제라고 볼 수 있다.

따라서, *precision*과 *recall*의 조화평균인 *f1-score*를 평가 지표로 선정하는 것이 타당하다.

이 경우 *f1-score*가 0.99로 가장 높은 **svm**이 가장 좋은 성능을 보여준다고 판단할 수 있다.



---



### **2. load_wine : 와인을 분류해 봅시다**

**(1) 필요한 모듈 import하기**

In [361]:
'''
import sklearn

print(sklearn.__version__)
'''

'\nimport sklearn\n\nprint(sklearn.__version__)\n'

In [362]:
from sklearn.datasets import load_wine
'''
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report
'''

'\nfrom sklearn.model_selection import train_test_split\nfrom sklearn.metrics import classification_report\n'

In [363]:
'''
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
'''
from sklearn import svm
'''
from sklearn.linear_model import SGDClassifier
from sklearn.linear_model import LogisticRegression
'''

'\nfrom sklearn.linear_model import SGDClassifier\nfrom sklearn.linear_model import LogisticRegression\n'

**(2) 데이터 준비**

In [364]:
wine = load_wine()

**(3) 데이터 이해하기**

*   Feature Data 지정하기

In [365]:
wine_feature = wine.data

*   Label Data 지정하기

In [366]:
wine_label = wine.target

*   Target Names 출력해 보기

In [367]:
wine.target_names

array(['class_0', 'class_1', 'class_2'], dtype='<U7')

*   데이터 Describe 해 보기

In [368]:
print(wine.DESCR)

.. _wine_dataset:

Wine recognition dataset
------------------------

**Data Set Characteristics:**

    :Number of Instances: 178 (50 in each of three classes)
    :Number of Attributes: 13 numeric, predictive attributes and the class
    :Attribute Information:
 		- Alcohol
 		- Malic acid
 		- Ash
		- Alcalinity of ash  
 		- Magnesium
		- Total phenols
 		- Flavanoids
 		- Nonflavanoid phenols
 		- Proanthocyanins
		- Color intensity
 		- Hue
 		- OD280/OD315 of diluted wines
 		- Proline

    - class:
            - class_0
            - class_1
            - class_2
		
    :Summary Statistics:
    
    ============================= ==== ===== ======= =====
                                   Min   Max   Mean     SD
    ============================= ==== ===== ======= =====
    Alcohol:                      11.0  14.8    13.0   0.8
    Malic Acid:                   0.74  5.80    2.34  1.12
    Ash:                          1.36  3.23    2.36  0.27
    Alcalinity of Ash:            1

**(4) train, test 데이터 분리**

In [369]:
X_train, X_test, y_train, y_test = train_test_split(wine_feature, wine_label, test_size=0.2, random_state=1)

**(5) 다양한 모델로 학습시켜보기**

*   Decision Tree 사용해 보기

In [370]:
decision_tree = DecisionTreeClassifier(random_state=1)
decision_tree.fit(X_train, y_train)
y_pred_decision_tree = decision_tree.predict(X_test)

print(classification_report(y_test, y_pred_decision_tree))

              precision    recall  f1-score   support

           0       0.93      0.93      0.93        14
           1       0.75      0.92      0.83        13
           2       1.00      0.67      0.80         9

    accuracy                           0.86        36
   macro avg       0.89      0.84      0.85        36
weighted avg       0.88      0.86      0.86        36



*   Random Forest 사용해 보기

In [371]:
random_forest = RandomForestClassifier(random_state=2)
random_forest.fit(X_train, y_train)
y_pred_random_forest = random_forest.predict(X_test)

print(classification_report(y_test, y_pred_random_forest))

              precision    recall  f1-score   support

           0       0.93      1.00      0.97        14
           1       1.00      0.92      0.96        13
           2       1.00      1.00      1.00         9

    accuracy                           0.97        36
   macro avg       0.98      0.97      0.98        36
weighted avg       0.97      0.97      0.97        36



* SVM 사용해 보기

In [372]:
svm = svm.SVC()
svm.fit(X_train, y_train)
y_pred_svm = svm.predict(X_test)

print(classification_report(y_test, y_pred_svm))

              precision    recall  f1-score   support

           0       0.92      0.79      0.85        14
           1       0.58      0.85      0.69        13
           2       0.20      0.11      0.14         9

    accuracy                           0.64        36
   macro avg       0.57      0.58      0.56        36
weighted avg       0.62      0.64      0.61        36



*   SGD Classifier 사용해 보기

In [373]:
sgd = SGDClassifier()
sgd.fit(X_train, y_train)
y_pred_sgd = sgd.predict(X_test)

print(classification_report(y_test, y_pred_sgd))

              precision    recall  f1-score   support

           0       1.00      0.86      0.92        14
           1       0.73      0.85      0.79        13
           2       0.56      0.56      0.56         9

    accuracy                           0.78        36
   macro avg       0.76      0.75      0.75        36
weighted avg       0.79      0.78      0.78        36



*   Logistic Regression 사용해 보기

In [374]:
logistic_regression = LogisticRegression(max_iter=3000)
logistic_regression.fit(X_train, y_train)
y_pred_logistic_regression = logistic_regression.predict(X_test)

print(classification_report(y_test, y_pred_logistic_regression))

              precision    recall  f1-score   support

           0       0.93      1.00      0.97        14
           1       0.92      0.92      0.92        13
           2       1.00      0.89      0.94         9

    accuracy                           0.94        36
   macro avg       0.95      0.94      0.94        36
weighted avg       0.95      0.94      0.94        36



**(6) 모델을 평가해 보기**

다른 조건들이 없다면 손글씨 데이터와 마찬가지로 *precision*과 *recall*은 큰 의미를 내포하고 있지 않다.

따라서, 동일하게 *f1-score*를 평가 지표로 선정하는 것이 타당하다.

이 경우 *f1-score*가 0.97로 가장 높은 **Random Forest**가 가장 좋은 성능을 보여준다고 판단할 수 있다.



---



### **3. load_breast_cancer : 유방암 여부를 진단해 봅시다**

**(1) 필요한 모듈 import하기**

In [375]:
'''
import sklearn

print(sklearn.__version__)
'''

'\nimport sklearn\n\nprint(sklearn.__version__)\n'

In [376]:
from sklearn.datasets import load_breast_cancer
'''
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report
'''

'\nfrom sklearn.model_selection import train_test_split\nfrom sklearn.metrics import classification_report\n'

In [377]:
'''
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
'''
from sklearn import svm
'''
from sklearn.linear_model import SGDClassifier
from sklearn.linear_model import LogisticRegression
'''

'\nfrom sklearn.linear_model import SGDClassifier\nfrom sklearn.linear_model import LogisticRegression\n'

**(2) 데이터 준비**

In [378]:
breast_cancer = load_breast_cancer()

**(3) 데이터 이해하기**

*   Feature Data 지정하기

In [379]:
breast_cancer_feature = breast_cancer.data

*   Label Data 지정하기

In [380]:
breast_cancer_label = breast_cancer.target

*   Target Names 출력해 보기

In [381]:
breast_cancer.target_names

array(['malignant', 'benign'], dtype='<U9')

*   데이터 Describe 해 보기

In [382]:
print(breast_cancer.DESCR)

.. _breast_cancer_dataset:

Breast cancer wisconsin (diagnostic) dataset
--------------------------------------------

**Data Set Characteristics:**

    :Number of Instances: 569

    :Number of Attributes: 30 numeric, predictive attributes and the class

    :Attribute Information:
        - radius (mean of distances from center to points on the perimeter)
        - texture (standard deviation of gray-scale values)
        - perimeter
        - area
        - smoothness (local variation in radius lengths)
        - compactness (perimeter^2 / area - 1.0)
        - concavity (severity of concave portions of the contour)
        - concave points (number of concave portions of the contour)
        - symmetry
        - fractal dimension ("coastline approximation" - 1)

        The mean, standard error, and "worst" or largest (mean of the three
        worst/largest values) of these features were computed for each image,
        resulting in 30 features.  For instance, field 0 is Mean Radi

**(4) train, test 데이터 분리**

In [383]:
X_train, X_test, y_train, y_test = train_test_split(breast_cancer_feature, breast_cancer_label, test_size=0.2, random_state=1)

**(5) 다양한 모델로 학습시켜보기**

*   Decision Tree 사용해 보기

In [384]:
decision_tree = DecisionTreeClassifier(random_state=1)
decision_tree.fit(X_train, y_train)
y_pred_decision_tree = decision_tree.predict(X_test)

print(classification_report(y_test, y_pred_decision_tree))

              precision    recall  f1-score   support

           0       0.97      0.88      0.93        42
           1       0.93      0.99      0.96        72

    accuracy                           0.95       114
   macro avg       0.95      0.93      0.94       114
weighted avg       0.95      0.95      0.95       114



*   Random Forest 사용해 보기

In [385]:
random_forest = RandomForestClassifier(random_state=2)
random_forest.fit(X_train, y_train)
y_pred_random_forest = random_forest.predict(X_test)

print(classification_report(y_test, y_pred_random_forest))

              precision    recall  f1-score   support

           0       0.97      0.88      0.93        42
           1       0.93      0.99      0.96        72

    accuracy                           0.95       114
   macro avg       0.95      0.93      0.94       114
weighted avg       0.95      0.95      0.95       114



* SVM 사용해 보기

In [386]:
svm = svm.SVC()
svm.fit(X_train, y_train)
y_pred_svm = svm.predict(X_test)

print(classification_report(y_test, y_pred_svm))

              precision    recall  f1-score   support

           0       1.00      0.74      0.85        42
           1       0.87      1.00      0.93        72

    accuracy                           0.90       114
   macro avg       0.93      0.87      0.89       114
weighted avg       0.92      0.90      0.90       114



*   SGD Classifier 사용해 보기

In [387]:
sgd = SGDClassifier()
sgd.fit(X_train, y_train)
y_pred_sgd = sgd.predict(X_test)

print(classification_report(y_test, y_pred_sgd))

              precision    recall  f1-score   support

           0       0.95      0.88      0.91        42
           1       0.93      0.97      0.95        72

    accuracy                           0.94       114
   macro avg       0.94      0.93      0.93       114
weighted avg       0.94      0.94      0.94       114



*   Logistic Regression 사용해 보기

In [388]:
logistic_regression = LogisticRegression(max_iter=3000)
logistic_regression.fit(X_train, y_train)
y_pred_logistic_regression = logistic_regression.predict(X_test)

print(classification_report(y_test, y_pred_logistic_regression))

              precision    recall  f1-score   support

           0       0.97      0.88      0.93        42
           1       0.93      0.99      0.96        72

    accuracy                           0.95       114
   macro avg       0.95      0.93      0.94       114
weighted avg       0.95      0.95      0.95       114



**(6) 모델을 평가해 보기**

앞선 데이터들과 달리 label 간 예측률이 높아야 하는 우선 순위가 존재한다.

즉, 모델은 '0'번 label인 **'malignant'(악성 종양)**에 대하여 훨씬 민감하게 반응해야 한다.

특히,
> 실제 **'malignant'(악성 종양)**를 'benign'(양성 종양)으로 예측하는 경우 최소화해야 할 것이다.

따라서 '0'번 label에 대한 *recall* 값을 평가 지표로 선정하는 것이 타당하다.

이 경우 해당 값이 0.90으로 가장 높은 **SGD Classifier**이 가장 좋은 성능을 보여준다고 판단할 수 있다.



---



### **4. 회고**

이번 노드를 통해 배운 점들은 크게 아래와 같다.

1.   싸이킷런 활용법
2.   머신러닝의 카테고리
3.   머신러닝 모델의 종류
4.   모델 성능 평가 지표

그리고 아쉬웠던 부분은, 모델 성능 평가 지표로서 끝내 **ROC 및 AUC**를 활용하지 못했던 것이다.

관련 메서드들을 이해하고 다음 기회에는 꼭 활용해보고 싶다.
